src/core/particle_data.cpp

```diff
@@ -1258,310 +1258,48 @@ void auto_exclusions(int distance) {
 
 #endif
 
-int init_gc(void) {
-  if (type_array == (TypeList *)nullptr) {
-    // stores the number of currently available type_list's
-    number_of_type_lists = 10;
-
-    Type.max_entry = 0;
-    Index.max_entry = 0;
+void init_type_map(int type) {
+  GC_init=true;
+  if (type < 0)
+    throw std::runtime_error("Types may not be negative");
 
-    type_array =
-        (TypeList *)Utils::malloc(sizeof(TypeList) * number_of_type_lists);
-    if (type_array == (TypeList *)0)
-      return ES_ERROR;
-
-    GC_init = 1;
-    Type_array_init = 0;
-  }
-  return ES_OK;
-}
-
-int init_type_array(int type) {
-  if (init_gc() == ES_ERROR)
-    return ES_ERROR;
-
-  for (int i = 0; i < Index.max_entry; i++)
-    if (type == Type.index[i] && Index.type[type] != -1) {
-      // already indexed
-      return ES_OK;
-    }
-
-  int type_index = -1;
-  type_index = (Type.max_entry++);
-  if (type_index == number_of_type_lists) {
-    reallocate_global_type_list(number_of_type_lists * 2);
-  }
-
-  Type.index = Utils::realloc(Type.index, sizeof(int) * Type.max_entry);
-
-  // reallocate the array that holds the particle type and points to the type
-  // index used for the type_list
-
-  if (type >= Index.max_entry) {
-    Index.type = Utils::realloc(Index.type, (type + 1) * sizeof(int));
-    Index.max_entry = type + 1;
-  }
-  for (int i = 0; i < Type.max_entry; i++)
-    Index.type[i] = -1;
-
-  if (Type.index == (int *)0 || Index.type == (int *)0)
-    return ES_ERROR;
-
-  // allocates a list for ids for as many entries as there are particles right
-  // now
-  if (type < 0) {
-    return ES_ERROR;
-  }
-  Type.index[type_index] = type;
-  // fill in array type_index_of_type
-  for (int i = 0; i < Type.max_entry; i++) {
-    Index.type[Type.index[i]] = i;
-  }
-
-  int t_c = 0; // index
-  type_array[Index.type[type]].id_list =
-      (int *)Utils::malloc(sizeof(int) * n_part);
+  //fill particle map
+  if(ParticleTypeMap.count(type)==0)
```

Comment:
This is only needed if you need the sets to exist for types that have no particles.
Otherwise you can just do
```c++
ParticleTypeMap[type].insert(id);
```

Reply:
yes because then the map will create a default constructed unordered set and will insert id into it

Reply:
Just use `ParticleTypeMap[type]` when getting the particle count.

Reply:
When using ParticleTypeMap[type] on an unitialized type this will result in a zero count then although the system might contain particles of the given type. In the current implementation 
int number_of_particles_with_type(int type) {
  return static_cast<int>(particle_type_map.at(type).size());
}
at least tells you that something went wrong since at checks for the type already existing in the particle_type_map. Here it is the users responsibility to initialize the set of ids for the types he is interested in.